### PR TITLE
Extending NodeListWithCursor

### PR DESCRIPTION
**Pull request by doctrino**
## Description
Please describe the change you have made.

## Checklist:
- [ ] Tests added/updated.
- [ ] Documentation updated. Documentation is generated from docstrings - these must be updated according to your change.
  If a new method has been added it should be referenced in [cognite.rst](https://github.com/cognitedata/cognite-sdk-python/blob/master/docs/source/cognite.rst) in order to generate docs based on its docstring.
- [ ] Changelog updated in [CHANGELOG.md](https://github.com/cognitedata/cognite-sdk-python/blob/master/CHANGELOG.md).
- [ ] Version bumped. If triggering a new release is desired, bump the version number in [_version.py](https://github.com/cognitedata/cognite-sdk-python/blob/master/cognite/client/_version.py) and [pyproject.toml](https://github.com/cognitedata/cognite-sdk-python/blob/master/pyproject.toml) per [semantic versioning](https://semver.org/).

### CODE REVIEW

File: tests/tests_unit/test_data_classes/test_data_models/test_instances.py [hunk 1]
```diff
@@ -161,6 +162,32 @@ def test_dump_and_load(self) -> None:
         }
 
 
+class TestNodeListWithCursor:
```
Comment (erlendvollset) [inline tests/tests_unit/test_data_classes/test_data_models/test_instances.py:165]: Add test for EdgeListWithCursor too?

File: cognite/client/data_classes/data_modeling/instances.py [hunk 2]
```diff
@@ -995,6 +1008,19 @@ def __init__(
...
 
+    def extend(self, other: EdgeListWithCursor) -> None:  # type: ignore[override]
+        if not isinstance(other, type(self)):
+            raise ValueError("Unable to extend as the types do not match")
+        other_res_list = type(self)(other, other.cursor)  # See if we can accept the types
+        if set(self._id_to_item).isdisjoint(other_res_list._id_to_item):
+            # Skip super() as we need the original type.
+            UserList.extend(self, other)
+            self._external_id_to_item.update(other_res_list._external_id_to_item)
+            self._id_to_item.update(other_res_list._id_to_item)
```
Comment (erlendvollset) [inline cognite/client/data_classes/data_modeling/instances.py:1019]: Edges and nodes don't have internal ids, so this is redundant.

File: cognite/client/data_classes/data_modeling/instances.py [hunk 3]
```diff
@@ -995,6 +1008,19 @@ def __init__(
         super().__init__(resources, cognite_client)
         self.cursor = cursor
 
+    def extend(self, other: EdgeListWithCursor) -> None:  # type: ignore[override]
+        if not isinstance(other, type(self)):
+            raise ValueError("Unable to extend as the types do not match")
+        other_res_list = type(self)(other, other.cursor)  # See if we can accept the types
+        if set(self._id_to_item).isdisjoint(other_res_list._id_to_item):
```
Comment (erlendvollset) [inline cognite/client/data_classes/data_modeling/instances.py:1015]: `_id_to_item` will always be empty, so this will always return `True`, need to use `_external_id_to_item`. Let's add unit tests for duplicates in extends.

Also 
```
set(self._external_id_to_item).isdisjoint(other_res_list._external_id_to_item)
```
is about 10 times slower than
```
self._external_id_to_item.keys().isdisjoint(other_res_list._external_id_to_item)
```

**Review by erlendvollset — CHANGES_REQUESTED** [review]

File: cognite/client/data_classes/data_modeling/instances.py (hunk 3 above)
Comment (doctrino) [reply cognite/client/data_classes/data_modeling/instances.py:1015]: Nice did not know that one, learned something new.
